### DISCUSSION TITLE
PWM discontinuties RP2

### DISCUSSION THREAD

**Discussion by karidavidsson** [micropython/micropython#9349] [General]
I am playing with PWM on the raspberry pico board.
When changing the PWM frequency the transition is not continuous, meaning that the duty cycle is not conserved with the new frequency.
![rp2_pwm_freq_change](https://user-images.githubusercontent.com/33786238/190926571-0f4e4ba4-9b35-4eec-b2cd-d704862528d8.png)
I take care to change the frequency of both PWM channels but as seen on the image both channels get some funky intermediate duty-cycle before the final correct duty cycle which I have to set again. 
The first trace changes the duty cycle from zero to 99% and the second trace should have fixed 50% duty-cycle on both frequencies.
What I would like to see is that the first trace has duty-cycle of 0 until duty-cycle of 99 is set (skip the 50% intermediary duty-cycle ). The second trace should have 50% duty-cycle the whole time and only the frequency change.
The code executes at the transition is :
```
def setoutput( pair, F, D ):
    pair[ 'output' ].freq( F )
    pair[ 'led' ].freq( F )

    pair[ 'led' ].duty_u16( D )
    pair[ 'output' ].duty_u16( 32767 )
```
Initialized by call of `setoutput( pair, 35103, 0 )`, followed by( `setouput( pair, 13603, 99)`. 
The `pair[ 'output' ]` is the second trace (brown) and `pair['led']` is the first trace (white). 
micropython is master as of Jul 15th 2022.

**Comment by scruss** [comment]
It's a quirk of the RP2040 that PWM duty cycle changes when you change frequency, and vice versa. You have to set both every time you change one.

**Comment by robert-hh** [comment]
@scruss If you change the frequency, the duty cycle will set internally again by the code for the freq() method to the initial state, as long as you work on a single channel.
@karidavidsson If you look at the properties of each of the two PWM pins in your test, which slice & channel number is shown?

**Reply by karidavidsson** [reply]
@robert-hh You are correct I had the the pins on different pwm slice. When using pins on the same pwm slice this looks much better.

![rp2_pwm](https://user-images.githubusercontent.com/33786238/191095016-e62631aa-5289-4055-b248-533b87feeb6c.png)
